File: Player/src/main/java/com/kaibo/music/player/PlayManager.java
package com.kaibo.music.player;

import android.app.Activity;
import android.content.ComponentName;
import android.content.Context;
import android.content.ContextWrapper;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.IBinder;
import android.os.RemoteException;

import com.kaibo.music.ISongService;
import com.kaibo.music.bean.SongBean;
import com.orhanobut.logger.Logger;

import java.util.ArrayList;
import java.util.List;
import java.util.WeakHashMap;

/**
 * Created by D22434 on 2017/9/20.
 */

public class PlayManager {
    public static ISongService mService = null;
    private static final WeakHashMap<Context, ServiceBinder> mConnectionMap;

    static {
        mConnectionMap = new WeakHashMap<>();
    }

    public static ServiceToken bindToService(Context context, ServiceConnection callback) {
        Activity realActivity = ((Activity) context).getParent();
        if (realActivity == null) {
            realActivity = (Activity) context;
        }
        final ContextWrapper contextWrapper = new ContextWrapper(realActivity);
        contextWrapper.startService(new Intent(contextWrapper, MusicPlayerService.class));
        final ServiceBinder binder = new ServiceBinder(callback);
        if (contextWrapper.bindService(new Intent().setClass(contextWrapper, MusicPlayerService.class), binder, 0)) {
            mConnectionMap.put(contextWrapper, binder);
            Logger.d("绑定成功");
            return new ServiceToken(contextWrapper);
        }else {
            Logger.d("绑定失败");
            return null;
        }
    }

    public static void unbindFromService(final ServiceToken token) {
        if (token == null) {
            return;
        }
        final ContextWrapper mContextWrapper = token.mWrappedContext;
        final ServiceBinder mBinder = mConnectionMap.get(mContextWrapper);
        if (mBinder == null) {
            return;
        }
        mContextWrapper.unbindService(mBinder);
        if (mConnectionMap.isEmpty()) {
            mService = null;
        }
    }

    public static final boolean isPlaybackServiceConnected() {
        return mService != null;
    }

    public static void nextPlay(SongBean music) {
        try {
            if (mService != null) {
                mService.nextPlay(music);
            }
        } catch (RemoteException e) {
            e.printStackTrace();
        }
    }

    public static void playOnline(SongBean music) {
        try {
            if (mService != null) {
                mService.playSong(music);
            }
        } catch (RemoteException e) {
            e.printStackTrace();
        }
    }

    public static void play(int id) {
        try {
            if (mService != null) {
                mService.play(id);
            }
        } catch (RemoteException e) {
            e.printStackTrace();
        }
    }

    public static void play(int id, List<SongBean> musicList, String pid) {
        try {
            if (mService != null) {
                mService.playPlaylist(musicList, id, pid);
            }
        } catch (RemoteException e) {
            e.printStackTrace();
        }
    }

    public static int getAudioSessionId() {
        try {
            if (mService != null) {
                return mService.getAudioSessionId();
            }
        } catch (RemoteException e) {
            e.printStackTrace();
        }
        return 0;
    }

    public static void playPause() {
        try {
            if (mService != null) {
                mService.playPause();
            }
        } catch (RemoteException e) {
            e.printStackTrace();
        }
    }

    public static void prev() {
        try {
            if (mService != null) {
                mService.prev();
            }
        } catch (RemoteException e) {
            e.printStackTrace();
        }
    }

    public static void next() {
        try {
            if (mService != null) {
                mService.next();
            }
        } catch (RemoteException e) {
            e.printStackTrace();
        }
    }

    public static void setPlayMode(int loopmode) {
        try {
            if (mService != null) {
                mService.setPlayMode(loopmode);
            }
        } catch (RemoteException e) {
            e.printStackTrace();
        }
    }

    public static void seekTo(int ms) {
        try {
            if (mService != null) {
                mService.seekTo(ms);
            }
        } catch (RemoteException e) {
            e.printStackTrace();
        }
    }

    public static int position() {
        try {
            if (mService != null) {
                return mService.position();
            }
        } catch (RemoteException e) {
            e.printStackTrace();
        }
        return 0;
    }

    public static int getCurrentPosition() {
        try {
            if (mService != null) {
                return mService.getCurrentPosition();
            }
        } catch (RemoteException e) {
            e.printStackTrace();
        }
        return 0;
    }

    public static int getDuration() {
        try {
            if (mService != null) {
                return mService.getDuration();
            }
        } catch (RemoteException e) {
            e.printStackTrace();
        }
        return 0;
    }

    public static String getSongName() {
        try {
            if (mService != null) {
                return mService.getSongName();
            }
        } catch (RemoteException e) {
            e.printStackTrace();
        }
        return "湖科音乐";
    }

    public static String getSongArtist() {
        try {
            if (mService != null) {
                return mService.getSongArtist();
            }
        } catch (RemoteException e) {
            e.printStackTrace();
        }
        return "湖科音乐";
    }

    public static boolean isPlaying() {
        try {
            if (mService != null) {
                return mService.isPlaying();
            }
        } catch (RemoteException e) {
            e.printStackTrace();
        }
        return false;
    }

    public static boolean isPause() {
        try {
            if (mService != null) {
                return mService.isPause();
            }
        } catch (RemoteException e) {
            e.printStackTrace();
        }
        return false;
    }

    public static SongBean getPlayingMusic() {
        try {
            if (mService != null) {
                return mService.getPlayingSongBean();
            }
        } catch (RemoteException e) {
            e.printStackTrace();
        }
        return null;
    }

    public static String getPlayingId() {
        try {
            if (mService != null && mService.getPlayingSongBean() != null) {
                return mService.getPlayingSongBean().getMid();
            }
        } catch (RemoteException e) {
            e.printStackTrace();
        }
        return "-1";
    }

    public static List<SongBean> getPlayList() {
        try {
            if (mService != null) {
                return mService.getPlayList();
            }
        } catch (RemoteException e) {
            e.printStackTrace();
        }
        return new ArrayList<>();
    }

    public static void setPlayList(List<SongBean> playlist) {

    }

    public static void clearQueue() {
        try {
            if (mService != null) {
                mService.clearQueue();
            }
        } catch (RemoteException e) {
            e.printStackTrace();
        }
    }

    public static void removeFromQueue(int adapterPosition) {
        try {
            if (mService != null) {
                mService.removeFromQueue(adapterPosition);
            }
        } catch (RemoteException e) {
            e.printStackTrace();
        }
    }

    public static void showDesktopLyric(boolean isShow) {
        try {
            if (mService != null) {
                mService.showDesktopLyric(isShow);
            }
        } catch (RemoteException e) {
            e.printStackTrace();
        }
    }

    public static final class ServiceBinder implements ServiceConnection {
        private final ServiceConnection mCallback;

        public ServiceBinder(final ServiceConnection callback) {
            mCallback = callback;
        }

        @Override
        public void onServiceConnected(final ComponentName className, final IBinder service) {
            // 获取到远端的Binder
            mService = ISongService.Stub.asInterface(service);
            if (mCallback != null) {
                mCallback.onServiceConnected(className, service);
            }
        }

        @Override
        public void onServiceDisconnected(final ComponentName className) {
            if (mCallback != null) {
                mCallback.onServiceDisconnected(className);
            }
            mService = null;
        }
    }

    public static final class ServiceToken {
        public ContextWrapper mWrappedContext;

        public ServiceToken(final ContextWrapper context) {
            mWrappedContext = context;
        }
    }
}

File: Player/src/main/java/com/kaibo/music/player/manager/PlayManager.kt
package com.kaibo.music.player.manager

import android.app.Activity
import android.content.*
import android.os.IBinder
import com.kaibo.music.ISongService
import com.kaibo.music.bean.SongBean
import com.kaibo.music.player.service.MusicPlayerService
import com.orhanobut.logger.Logger
import java.util.*

/**
 * 对外提供的播放控制
 */
object PlayManager {

    private var mService: ISongService? = null

    private val mConnectionMap = WeakHashMap<Context, ServiceBinder>()

    /**
     * 获取当前的所播放歌曲的播放进度
     */
    val currentPosition: Int
        get() {
            return try {
                mService?.currentPosition ?: 0
            } catch (e: Exception) {
                e.printStackTrace()
                0
            }
        }

    /**
     * 获取当前所播放的歌曲的总时长
     */
    val duration: Int
        get() {
            return try {
                mService?.duration ?: 0
            } catch (e: Exception) {
                e.printStackTrace()
                0
            }
        }

    /**
     * 是否正在播放
     */
    val isPlaying: Boolean
        get() {
            return try {
                mService?.isPlaying ?: false
            } catch (e: Exception) {
                e.printStackTrace()
                false
            }
        }

    /**
     * 获取正在播放的歌曲  没有播放返回null
     *
     * @return
     */
    val playSong: SongBean?
        get() {
            return try {
                mService?.playSong
            } catch (e: Exception) {
                e.printStackTrace()
                null
            }
        }

    /**
     * 获取当前正在播放的播放队列
     */
    val playSongQueue: List<SongBean>
        get() {
            return try {
                mService?.playSongQueue ?: emptyList()
            } catch (e: Exception) {
                e.printStackTrace()
                emptyList()
            }
        }

    /**
     * 设置一首歌曲到播放队列  并执行播放
     */
    fun setPlaySong(songBean: SongBean) {
        try {
            mService?.playSong = songBean
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    /**
     * 设置一首歌曲到播放队列  但在下一次切歌的时候才执行播放
     */
    fun nextPlay(songBean: SongBean) {
        try {
            mService?.setNextSong(songBean)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    /**
     * 播放位置
     */
    var playPosition: Int
        set(value) {
            try {
                mService?.playPosition = value
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        get() {
            return try {
                mService?.playPosition ?: -1
            } catch (e: Exception) {
                e.printStackTrace()
                -1
            }
        }

    /**
     * 指定一个播放列表开始播放
     */
    fun setPlaySongList(playlist: List<SongBean>, position: Int, playListId: String) {
        try {
            mService?.setPlaySongList(playlist, position, playListId)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    /**
     * 切换播放暂停
     */
    fun togglePlayer() {
        try {
            mService?.togglePlayer()
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    /**
     * 上一曲
     */
    fun prev() {
        try {
            mService?.prev()
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    /**
     * 下一曲
     */
    fun next() {
        try {
            mService?.next()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun seekTo(pos: Int) {
        try {
            mService?.seekTo(pos)
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    fun clearQueue() {
        try {
            mService?.clearQueue()
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    fun removeFromQueue(adapterPosition: Int) {
        try {
            mService?.removeFromQueue(adapterPosition)
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    fun showDesktopLyric(isShow: Boolean) {
        try {
            mService?.showDesktopLyric(isShow)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    /**
     * 切换播放模式
     */
    fun updatePlayMode(): String {
        return try {
            mService?.updatePlayMode() ?: "切换失败"
        } catch (e: Exception) {
            e.printStackTrace()
            "切换失败"
        }
    }

    /**
     * 绑定播放Service到指定的Activity
     */
    fun bindToService(context: Context): ServiceToken? {
        var realActivity: Activity? = (context as Activity).parent
        if (realActivity == null) {
            realActivity = context
        }
        val contextWrapper = ContextWrapper(realActivity)
        // 启动Service
        contextWrapper.startService(Intent(contextWrapper, MusicPlayerService::class.java))
        val binder = ServiceBinder()
        // 绑定Service
        return if (contextWrapper.bindService(Intent().setClass(contextWrapper, MusicPlayerService::class.java), binder, 0)) {
            mConnectionMap[contextWrapper] = binder
            Logger.d("绑定成功")
            ServiceToken(contextWrapper)
        } else {
            Logger.e("绑定失败")
            null
        }
    }

    /**
     * 解除绑定
     */
    fun unbindFromService(token: ServiceToken?) {
        val mContextWrapper = token?.mWrappedContext ?: return
        val mBinder = mConnectionMap[mContextWrapper] ?: return
        mContextWrapper.unbindService(mBinder)
        if (mConnectionMap.isEmpty()) {
            mService = null
        }
    }

    class ServiceBinder : ServiceConnection {
        override fun onServiceConnected(className: ComponentName, service: IBinder) {
            // 获取到远端的Binder
            mService = ISongService.Stub.asInterface(service)
        }

        override fun onServiceDisconnected(className: ComponentName) {
            mService = null
        }
    }

    class ServiceToken(var mWrappedContext: ContextWrapper)
}
